source/exec/cache.d

```diff
@@ -44,17 +41,32 @@ class Cache: IExecProvider
 		import std.algorithm: canFind;
 		auto hash = getSourceCodeHash(source);
 
-		if (!assumeSorted(this.allowedSources_).canFind(hash)) {
-			auto result = realExecProvider_.compileAndExecute(source);
-			return result;
-		} else if (auto cache = hash in sourceHashToOutput_) {
-			return *cache;
+		if (auto cache = hash in sourceHashToOutput_) {
+			cache.time = Clock.currTime;
+			return cache.source;
 		} else {
 			auto result = realExecProvider_.compileAndExecute(source);
-			sourceHashToOutput_[hash] = result;
+			testCacheSize();
+			sourceHashToOutput_[hash] = ResultTuple(result, Clock.currTime);
 			return result;
 		}
 	}
+
+	void testCacheSize()
+	{
+		import std.stdio;
+		import std.array : array, byPair;
+		if (sourceHashToOutput_.length > maximumCacheSize)
+		{
+			// throw away the older half
+			foreach (entry; sourceHashToOutput_
+								.byPair
+								.map!(a => tuple!("key", "time")(a[0], a[1].time))
+								.array
+								.sort!((a, b) => a.time < b.time)[0 .. $ / 2])
+				sourceHashToOutput_.remove(entry.key);
```

Comment:
Depending on the implementation it might be faster to alloc a new dictionary and insert the new "half" though I doubt this (and it shouldn't matter anyway because this is very infrequently called).